### ISSUE TITLE
OpenMP 16.0.0-rc1 Build failure for arm

### ISSUE THREAD

**Issue by satmandu** [llvm/llvm-project#60487]
Using the same build setup as in https://github.com/llvm/llvm-project/issues/60393 (which succeeds on x86_64 and i686 with the patch from https://reviews.llvm.org/D143200 applied), building with either Clang 16.0.0-rc1 or Clang 15.0.7, in the ChromeOS/Chromebrew environment I have the error below.

This was resolved in openmp 15.0.x if I modified `openmp/CMakeLists.txt` to add `-fPIC` to `set(CMAKE_C_FLAGS` and `set(CMAKE_CXX_FLAGS`, but that no longer solves the issue.

```
[34/34] Linking C shared library runtime/src/libomp.so
FAILED: runtime/src/libomp.so
: && /usr/local/bin/clang --target=armv7l-cros-linux-gnueabihf -fPIC -O2 -pipe -ffat-lto-objects -fPIC -mfloat-abi=hard -mthumb -mfpu=vfpv3-d16 -march=armv7-a+fp -fuse-ld=mold -O3 -DNDEBUG  -Wl,--as-needed -static-libgcc -Wl,-z,noexecstack -shared -Wl,-soname,libomp.so -o runtime/src/libomp.so runtime/src/CMakeFiles/omp.dir/kmp_alloc.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_atomic.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_csupport.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_debug.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_itt.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_environment.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_error.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_global.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_i18n.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_io.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_runtime.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_settings.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_str.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_tasking.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_threadprivate.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_utility.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_barrier.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_wait_release.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_affinity.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_dispatch.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_lock.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_sched.cpp.o runtime/src/CMakeFiles/omp.dir/z_Linux_util.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_gsupport.cpp.o runtime/src/CMakeFiles/omp.dir/thirdparty/ittnotify/ittnotify_static.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_taskdeps.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_cancel.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_ftn_cdecl.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_ftn_extra.cpp.o runtime/src/CMakeFiles/omp.dir/kmp_version.cpp.o runtime/src/CMakeFiles/omp.dir/z_Linux_asm.S.o  -lpthread  -lrt  -lm  -ldl && cd /usr/local/tmp/crew/openmp.20230202185215.dir/openmp/builddir/runtime/src && /usr/local/bin/cmake -E create_symlink libomp.so libgomp.so && /usr/local/bin/cmake -E create_symlink libomp.so libiomp5.so
mold: error: runtime/src/CMakeFiles/omp.dir/kmp_ftn_cdecl.cpp.o:(.ARM.extab): R_ARM_PREL31 relocation at offset 0x0 against symbol `__gxx_personality_v0' can not be used; recompile with -fPIC
mold: error: runtime/src/CMakeFiles/omp.dir/kmp_ftn_extra.cpp.o:(.ARM.extab): R_ARM_PREL31 relocation at offset 0x0 against symbol `__gxx_personality_v0' can not be used; recompile with -fPIC
mold: error: runtime/src/CMakeFiles/omp.dir/kmp_ftn_extra.cpp.o:(.ARM.extab): R_ARM_PREL31 relocation at offset 0x1c against symbol `__gxx_personality_v0' can not be used; recompile with -fPIC
mold: error: runtime/src/CMakeFiles/omp.dir/kmp_ftn_extra.cpp.o:(.ARM.extab): R_ARM_PREL31 relocation at offset 0x44 against symbol `__gxx_personality_v0' can not be used; recompile with -fPIC
mold: error: runtime/src/CMakeFiles/omp.dir/kmp_affinity.cpp.o:(.ARM.extab): R_ARM_PREL31 relocation at offset 0xc against symbol `__gxx_personality_v0' can not be used; recompile with -fPIC
mold: error: runtime/src/CMakeFiles/omp.dir/kmp_affinity.cpp.o:(.ARM.extab): R_ARM_PREL31 relocation at offset 0x24 against symbol `__gxx_personality_v0' can not be used; recompile with -fPIC
mold: error: runtime/src/CMakeFiles/omp.dir/kmp_affinity.cpp.o:(.ARM.extab): R_ARM_PREL31 relocation at offset 0x3c against symbol `__gxx_personality_v0' can not be used; recompile with -fPIC
mold: error: runtime/src/CMakeFiles/omp.dir/kmp_affinity.cpp.o:(.ARM.extab): R_ARM_PREL31 relocation at offset 0x54 against symbol `__gxx_personality_v0' can not be used; recompile with -fPIC
mold: error: runtime/src/CMakeFiles/omp.dir/kmp_affinity.cpp.o:(.ARM.extab): R_ARM_PREL31 relocation at offset 0xa0 against symbol `__gxx_personality_v0' can not be used; recompile with -fPIC
mold: error: runtime/src/CMakeFiles/omp.dir/kmp_affinity.cpp.o:(.ARM.extab.text._ZN17KMPNativeAffinityD0Ev): R_ARM_PREL31 relocation at offset 0x0 against symbol `__gxx_personality_v0' can not be used; recompile with -fPIC
mold: error: runtime/src/CMakeFiles/omp.dir/kmp_affinity.cpp.o:(.ARM.extab.text._ZN17KMPNativeAffinity13allocate_maskEv): R_ARM_PREL31 relocation at offset 0x0 against symbol `__gxx_personality_v0' can not be used; recompile with -fPIC
mold: error: runtime/src/CMakeFiles/omp.dir/kmp_affinity.cpp.o:(.ARM.extab.text._ZN17KMPNativeAffinity19allocate_mask_arrayEi): R_ARM_PREL31 relocation at offset 0x0 against symbol `__gxx_personality_v0' can not be used; recompile with -fPIC
mold: error: runtime/src/CMakeFiles/omp.dir/kmp_affinity.cpp.o:(.ARM.extab.text._ZN17KMPNativeAffinity21deallocate_mask_arrayEPN11KMPAffinity4MaskE): R_ARM_PREL31 relocation at offset 0x0 against symbol `__gxx_personality_v0' can not be used; recompile with -fPIC
mold: error: runtime/src/CMakeFiles/omp.dir/kmp_affinity.cpp.o:(.ARM.extab.text._ZN17KMPNativeAffinity4MaskD2Ev): R_ARM_PREL31 relocation at offset 0x0 against symbol `__gxx_personality_v0' can not be used; recompile with -fPIC
mold: error: runtime/src/CMakeFiles/omp.dir/kmp_affinity.cpp.o:(.ARM.extab.text._ZN17KMPNativeAffinity4MaskD0Ev): R_ARM_PREL31 relocation at offset 0x0 against symbol `__gxx_personality_v0' can not be used; recompile with -fPIC
mold: error: runtime/src/CMakeFiles/omp.dir/kmp_barrier.cpp.o:(.ARM.extab.text._ZN15kmp_flag_oncoreD0Ev): R_ARM_PREL31 relocation at offset 0x0 against symbol `__gxx_personality_v0' can not be used; recompile with -fPIC
mold: error: runtime/src/CMakeFiles/omp.dir/kmp_ftn_cdecl.cpp.o:(.ARM.extab): R_ARM_PREL31 relocation at offset 0x1c against symbol `__gxx_personality_v0' can not be used; recompile with -fPIC
mold: error: runtime/src/CMakeFiles/omp.dir/kmp_ftn_cdecl.cpp.o:(.ARM.extab): R_ARM_PREL31 relocation at offset 0x44 against symbol `__gxx_personality_v0' can not be used; recompile with -fPIC
mold: error: runtime/src/CMakeFiles/omp.dir/kmp_runtime.cpp.o:(.ARM.extab.text._ZN11kmp_flag_64ILb0ELb1EED0Ev): R_ARM_PREL31 relocation at offset 0x0 against symbol `__gxx_personality_v0' can not be used; recompile with -fPIC
mold: error: runtime/src/CMakeFiles/omp.dir/kmp_barrier.cpp.o:(.ARM.extab.text._ZN11kmp_flag_64ILb1ELb0EED0Ev): R_ARM_PREL31 relocation at offset 0x0 against symbol `__gxx_personality_v0' can not be used; recompile with -fPIC
mold: error: runtime/src/CMakeFiles/omp.dir/kmp_ftn_cdecl.cpp.o:(.ARM.extab.text._ZN15ConvertedStringD2Ev): R_ARM_PREL31 relocation at offset 0x0 against symbol `__gxx_personality_v0' can not be used; recompile with -fPIC
clang-16: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

[CMakeError.log](https://github.com/llvm/llvm-project/files/10572499/CMakeError.log)
[CMakeOutput.log](https://github.com/llvm/llvm-project/files/10572500/CMakeOutput.log)
[openmp-16.0.0-rc1-armv7l.log](https://github.com/llvm/llvm-project/files/10572502/openmp-16.0.0-rc1-armv7l.log)

**Comment by jpeyton52** [comment]
All the link errors relate to `__gxx_personality_v0` (C++ exceptions mechanism) which implies the -fno-exceptions and -fno-rtti flags weren't included in the compilations of the source files. The flags aren't included due to the same issue as in #60393, and I can verify this from the CMakeError.log output.

```
Performing C++ SOURCE FILE Test LIBOMP_HAVE_FNO_EXCEPTIONS_FLAG failed with the following output:
Change Dir: /usr/local/tmp/crew/openmp.20230202172248.dir/openmp/builddir/CMakeFiles/CMakeScratch/TryCompile-Qe42Nu

Run Build Command(s):/usr/local/bin/ninja cmTC_f1b30 && [1/2] Building CXX object CMakeFiles/cmTC_f1b30.dir/src.cxx.o
clang-15: warning: optimization flag '-ffat-lto-objects' is not supported [-Wignored-optimization-argument]
clang-15: warning: argument unused during compilation: '-fuse-ld=mold' [-Wunused-command-line-argument]
[2/2] Linking CXX executable cmTC_f1b30


Source file was:
int main() { return 0; }

Performing C++ SOURCE FILE Test LIBOMP_HAVE_FNO_RTTI_FLAG failed with the following output:
Change Dir: /usr/local/tmp/crew/openmp.20230202172248.dir/openmp/builddir/CMakeFiles/CMakeScratch/TryCompile-P8FHkV

Run Build Command(s):/usr/local/bin/ninja cmTC_6c801 && [1/2] Building CXX object CMakeFiles/cmTC_6c801.dir/src.cxx.o
clang-15: warning: optimization flag '-ffat-lto-objects' is not supported [-Wignored-optimization-argument]
clang-15: warning: argument unused during compilation: '-fuse-ld=mold' [-Wunused-command-line-argument]
[2/2] Linking CXX executable cmTC_6c801


Source file was:
int main() { return 0; }
```

Can you remove the -ffat-lto-objects flag and see what happens?

**Comment by satmandu** [comment]
Huh. removing that fixes the build...
